File: lib/speech.js
/**
 * TTS via the browser's Web Speech API — no server proxy, no API keys.
 *
 * Mandarin is the main use (word cards, example sentences), but the quiz can
 * also read a *meaning* out loud, which has to use the UI language — so the
 * voice picker is parameterized by locale instead of being zh-only.
 *
 * Voice quality: browsers often default to a robotic local voice, so the
 * candidates are scored and the most natural one wins: remote/cloud voices
 * (Google 普通话, Microsoft Online Natural) beat local ones, and
 * "natural/neural/premium/enhanced" variants beat plain defaults.
 */

/** UI language code → preferred BCP-47 locale for synthesis. */
const LOCALE_BY_LANG = {
  zh: 'zh-CN',
  ru: 'ru-RU',
  en: 'en-US',
  tk: 'tk-TM',
}

/** Mandarin is read slightly slower than normal; prose at (near) full speed. */
const RATE_BY_LANG = { zh: 0.85 }

// Resolved voice per locale ('zh-CN' → SpeechSynthesisVoice | null).
const voiceCache = new Map()

const normalizeLocale = (value) => String(value || '').replace('_', '-').toLowerCase()

/**
 * Scores a voice against a target locale. Returns -1 for a different language.
 * Mandarin gets the extra `cmn-*` aliases some engines report.
 */
function scoreVoice(voice, target) {
  const lang = normalizeLocale(voice.lang)
  const name = String(voice.name || '').toLowerCase()
  const wanted = normalizeLocale(target)
  const primary = wanted.split('-')[0]
  const isChinese = primary === 'zh'

  let score = 0
  if (lang === wanted || (isChinese && lang === 'cmn-cn')) {
    score += 10 // exact region match (zh-CN, ru-RU, …)
  } else if (lang.split('-')[0] === primary || (isChinese && lang.startsWith('cmn'))) {
    score += 4 // same language, different region (zh-TW, ru-BY, …)
  } else {
    return -1
  }

  // Cloud voices are the near-native ones (Chrome: localService=false).
  if (voice.localService === false) score += 8
  if (/natural|neural/.test(name)) score += 6 // Edge "Xiaoxiao Online (Natural)"
  if (/premium|enhanced/.test(name)) score += 4 // iOS/macOS enhanced voices
  if (/google/.test(name)) score += 3 // "Google 普通话（中国大陆）"
  if (/siri/.test(name)) score += 2
  return score
}

function pickVoice(target) {
  if (voiceCache.has(target)) return voiceCache.get(target)
  const voices = window.speechSynthesis?.getVoices?.() || []
  let best = null
  let bestScore = 0
  for (const voice of voices) {
    const score = scoreVoice(voice, target)
    if (score > bestScore) {
      bestScore = score
      best = voice
    }
  }
  // An empty voice list means the engine has not loaded them yet — don't cache
  // that as "no voice", or the first utterance would poison every later one.
  if (voices.length) voiceCache.set(target, best)
  return best
}

export function canSpeak() {
  return typeof window !== 'undefined' && 'speechSynthesis' in window
}

/**
 * Speaks `text` in the given UI language ('zh' | 'ru' | 'en' | 'tk'),
 * cancelling any ongoing utterance. Returns false when speech is unavailable.
 *
 * A missing voice for the language is not a hard failure: the utterance still
 * carries the locale, so the engine can fall back to whatever it has (relevant
 * for Turkmen, which most platforms don't ship).
 */
export function speak(text, { lang = 'zh', rate } = {}) {
  if (!canSpeak() || !text) return false
  const locale = LOCALE_BY_LANG[lang] || LOCALE_BY_LANG.en
  const synth = window.speechSynthesis
  synth.cancel()
  const utterance = new SpeechSynthesisUtterance(text)
  utterance.lang = locale
  utterance.rate = rate ?? RATE_BY_LANG[lang] ?? 1
  utterance.pitch = 1
  const voice = pickVoice(locale)
  if (voice) utterance.voice = voice
  synth.speak(utterance)
  return true
}

export function speakChinese(text, { rate = RATE_BY_LANG.zh } = {}) {
  return speak(text, { lang: 'zh', rate })
}

/**
 * Long-form narration (Books AI retelling): splits `text` into sentences and
 * chains utterances so engines with per-utterance length caps don't cut off
 * mid-story. Returns a controller { stop() } or null when speech is
 * unavailable; `onDone` fires after the last sentence or on stop().
 */
export function splitSpeechText(text, maxLength = 180) {
  const limit = Math.max(1, Math.floor(maxLength) || 180)
  return String(text).split(/(?<=[。！？；!?;\n])/u).flatMap((sentence) => {
    const characters = Array.from(sentence.trim())
    const chunks = []
    for (let start = 0; start < characters.length; start += limit) {
      chunks.push(characters.slice(start, start + limit).join(''))
    }
    return chunks
  })
}

export function speakLong(text, { lang = 'zh', rate, onDone, onError } = {}) {
  if (!canSpeak() || !text) return null
  const locale = LOCALE_BY_LANG[lang] || LOCALE_BY_LANG.en
  const synth = window.speechSynthesis
  synth.cancel()

  const parts = splitSpeechText(text)
  if (parts.length === 0) return null

  let stopped = false
  let index = 0
  const finish = () => {
    if (stopped) return
    stopped = true
    onDone?.()
  }
  const speakNext = () => {
    if (stopped) return
    if (index >= parts.length) {
      finish()
      return
    }
    const utterance = new SpeechSynthesisUtterance(parts[index])
    index += 1
    utterance.lang = locale
    utterance.rate = rate ?? RATE_BY_LANG[lang] ?? 1
    utterance.pitch = 1
    const voice = pickVoice(locale)
    if (voice) utterance.voice = voice
    utterance.onend = speakNext
    utterance.onerror = (event) => {
      if (stopped) return
      if (!['canceled', 'interrupted'].includes(event.error)) onError?.(event.error)
      finish()
    }
    try { synth.speak(utterance) } catch (error) { onError?.(error); finish() }
  }
  speakNext()

  return {
    stop() {
      finish()
      synth.cancel()
    },
  }
}

// Voice list loads asynchronously in some browsers.
if (typeof window !== 'undefined' && 'speechSynthesis' in window) {
  window.speechSynthesis.onvoiceschanged = () => {
    voiceCache.clear()
  }
}

File: mobile/lib/core/speech.dart
import 'dart:io' show Platform;

import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:flutter_tts/flutter_tts.dart';

/// TTS via the platform speech engine — Flutter port of the web `lib/speech.js`.
///
/// Mandarin is the main use (word cards, example sentences), but the quiz can
/// also read a *meaning* out loud, which has to use the UI language — so the
/// engine is configured per language instead of being zh-only. Every new
/// utterance cancels the previous one.
///
/// Voice quality: instead of the engine's default (often a robotic local
/// voice), the best available voice for the target locale is picked explicitly:
///  1. the Google speech engine is preferred on Android when installed
///     (`com.google.android.tts` — noticeably closer to a native speaker
///     than most vendor engines);
///  2. among that locale's voices, network voices beat local ones and
///     higher `quality`/`x-...` variants beat plain defaults (Android voice
///     names look like `cmn-cn-x-ssa-network`).
class Speech {
  Speech._();

  static final FlutterTts _tts = FlutterTts();
  static Future<bool>? _engineReady;

  static const String _googleEngine = 'com.google.android.tts';

  /// UI language code → preferred BCP-47 locale, with fallbacks in order.
  static const Map<String, List<String>> _localesByLang = {
    'zh': ['zh-CN', 'zh'],
    'ru': ['ru-RU', 'ru'],
    'en': ['en-US', 'en'],
    'tk': ['tk-TM', 'tk'],
  };

  /// flutter_tts treats 0.5 as the normal rate. The web reads Mandarin at 0.85
  /// of normal (0.5 × 0.85 ≈ 0.42); prose runs at the normal rate.
  static const Map<String, double> _rateByLang = {'zh': 0.42};

  /// Resolved locale per language; the key is present once resolution ran,
  /// with a null value meaning "this device cannot speak that language".
  static final Map<String, String?> _resolvedLocale = {};
  static final Map<String, Map<String, String>?> _voiceByLocale = {};

  /// Language the engine is currently configured for — avoids re-issuing
  /// setLanguage/setVoice/setSpeechRate for consecutive utterances.
  static String? _currentLang;

  /// One-time engine selection. Resolves to false when the platform has no
  /// usable TTS engine (speak then no-ops, mirroring the web `canSpeak()`).
  static Future<bool> _ensureEngine() {
    return _engineReady ??= () async {
      try {
        await _pickEngine();
        // Neutral pitch — pitch shifts make Mandarin tones sound off.
        await _tts.setPitch(1.0);
        return true;
      } catch (_) {
        return false;
      }
    }();
  }

  /// Switch to the Google TTS engine on Android when it is installed and not
  /// already the default. Best-effort: any failure keeps the default engine.
  static Future<void> _pickEngine() async {
    if (kIsWeb || !Platform.isAndroid) return;
    try {
      final engines = await _tts.getEngines;
      if (engines is! List) return;
      final ids = [for (final e in engines) e.toString()];
      if (!ids.contains(_googleEngine)) return;
      final current = await _tts.getDefaultEngine;
      if (current?.toString() == _googleEngine) return;
      await _tts.setEngine(_googleEngine);
    } catch (_) {
      // Keep whatever engine the device defaults to.
    }
  }

  /// First locale from [_localesByLang] the engine reports as available, or
  /// null when it can speak none of them (e.g. Turkmen on most devices).
  static Future<String?> _resolveLocale(String lang) async {
    if (_resolvedLocale.containsKey(lang)) return _resolvedLocale[lang];
    String? found;
    for (final locale in _localesByLang[lang] ?? const <String>[]) {
      try {
        if (await _tts.isLanguageAvailable(locale) == true) {
          found = locale;
          break;
        }
      } catch (_) {
        // Engine cannot answer — treat as unavailable and try the next one.
      }
    }
    _resolvedLocale[lang] = found;
    return found;
  }

  /// Points the engine at [lang]. Returns false when the device has no voice
  /// for it, so the caller stays silent instead of reading text with a
  /// wrong-language voice.
  static Future<bool> _applyLanguage(String lang) async {
    if (_currentLang == lang) return true;
    final locale = await _resolveLocale(lang);
    if (locale == null) return false;
    try {
      await _tts.setLanguage(locale);
      final voice = await _pickBestVoice(locale);
      if (voice != null) await _tts.setVoice(voice);
      await _tts.setSpeechRate(_rateByLang[lang] ?? 0.5);
      _currentLang = lang;
      return true;
    } catch (_) {
      return false;
    }
  }

  /// Score-and-pick the most natural voice the engine offers for [locale].
  static Future<Map<String, String>?> _pickBestVoice(String locale) async {
    if (_voiceByLocale.containsKey(locale)) return _voiceByLocale[locale];

    Map<String, String>? best;
    try {
      final voices = await _tts.getVoices;
      // Voice listing unsupported — the engine default is still usable, and
      // this is not cached as a decision since the list may appear later.
      if (voices is! List) return null;

      final wanted = locale.toLowerCase().replaceAll('_', '-');
      final primary = wanted.split('-').first;
      final isChinese = primary == 'zh';
      var bestScore = -1;

      for (final raw in voices) {
        if (raw is! Map) continue;
        final name = raw['name']?.toString() ?? '';
        final voiceLocale = raw['locale']?.toString() ?? '';
        if (name.isEmpty || voiceLocale.isEmpty) continue;

        final loc = voiceLocale.toLowerCase().replaceAll('_', '-');
        var score = 0;
        // Mandarin (mainland) also shows up as cmn-cn on some engines. Other
        // zh variants (Cantonese/Taiwanese) only count as a weak fallback so
        // the pronunciation matches the pinyin shown on the cards.
        if (loc == wanted || (isChinese && loc == 'cmn-cn')) {
          score += 10;
        } else if (loc.split('-').first == primary ||
            (isChinese && loc.startsWith('cmn'))) {
          score += 4;
        } else {
          continue;
        }

        final n = name.toLowerCase();
        // Network voices are the neural, near-native ones.
        if (n.contains('network')) score += 8;
        // Enhanced/premium local variants still beat the plain default.
        if (n.contains('enhanced') || n.contains('premium')) score += 4;
        if (n.contains('-x-')) score += 2; // named Google voice families

        if (score > bestScore) {
          bestScore = score;
          best = {'name': name, 'locale': voiceLocale};
        }
      }
      _voiceByLocale[locale] = best;
    } catch (_) {
      return null;
    }
    return best;
  }

  /// Speaks [text] in [lang] ('zh' | 'ru' | 'en' | 'tk'), cancelling any
  /// ongoing utterance. Returns false when the device cannot speak it.
  static Future<bool> speak(String text, {String lang = 'zh'}) async {
    if (text.isEmpty) return false;
    if (!await _ensureEngine()) return false;
    final target = _localesByLang.containsKey(lang) ? lang : 'en';
    if (!await _applyLanguage(target)) return false;
    try {
      await _tts.stop();
      await _tts.speak(text);
      return true;
    } catch (_) {
      return false;
    }
  }

  /// Speak [text] with the Mandarin voice (same contract as the web
  /// `speakChinese`).
  static Future<bool> speakChinese(String text) => speak(text, lang: 'zh');
}
